### DISCUSSION TITLE
FLTK for Lua on windows

### DISCUSSION THREAD

**Discussion by DiegoJArg** [keplerproject/luarocks#1465] [General]
- **Platform:** Windows 10
- **LuaRocks version:** `luarocks --version` = 3.9.1 'Lua -v' = 5.1
- **Description:**

I am noob with Lua/Luarocks, and having difficulties to accomplish a test with FLTK.

I picked FLTK just to learn Lua and compare with pyFLTK.
And to match a lightweight interpreter with a lightweight UI.

## This was my battle:

- Luaforwindows, the bundle referenced at the oficial website, brings lua5.1 with luarocks 2.0.2.
- I tried to instal fltk4lua by doing `luarocks install fltk4lua`, just like one does `pip install pyfltk`.
- But it brings me to this error result shown below, which I found is because of an outdated version of luarocks.
```
.......\Lua\5.1\lua/luarocks\fs\lua.lua:526: attempt to concatenate local 'err' (a nil value) 
stack traceback:
        ...\Program Files (x86)\Lua\5.1\lua/luarocks\fs\lua.lua:526: in function 'download'
        ...:\Program Files (x86)\Lua\5.1\lua/luarocks\fetch.lua:33: in function 'fetch_url'
......
```
- Then unziped a new version (3.9.1) on the Luaforwindows folder.
- But luarocks is not automatically compatible, it comes prepared to work with lua5.4, not officially available for windows as binary.
- As luarocks doesn't detects the lua version by itself, I specified luarocks --lua-version=5.1 install fltk4lua
- Just then luarocks391 will work with bundled lua5.1. 
   - I have also tested with non-bundle binaries for lua5.4 from https://joedf.ahkscript.org/LuaBuilds

### Now the struggle continues about FLTK for Lua.

- I downloaded FLTK binaries for windows, but this wasn't enough, because I noted that fltk4lua is source-packaged and doesn't contains binaries of fltk:
    - https://code.google.com/archive/p/fltkwinbin/
- Since FLTK2 has "compat" headers for FL folder, I realised that fltk4lua is made for old FLTK 1.3 and it was detected correctly with `luarocks391 --lua-version=5.1 install fltk4lua FLTK_DIR=C:\Users\Note13\Downloads\fltk-1.3.x-r8627-win-bin`
- Still Missing dependencies for fltk4lua 0.2-1: wont install
    - luarocks-fetch-gitrec (not installed) 
    - luarocks-build-cpp (not installed)
- Trying to install both, produces the same output:
    - `No existing manifest. Attempting to rebuild...` 
    - and terminates without any error but it doesn't gets installed.

At this moment, I am realising after 2 days of struggling, that Lua requires to be closer to a compiler like MinGW, C/C++ headers, static/dinamic linking, etc.  I wasn't expecting this. pyFLTK probably already contains FLTK binaries bundled.

- **Is there anything else that I could do to accomplish FLTK over Lua without getting into compiler stuff just yet?**

**Comment by clitic** [comment]
@DiegoJArg You can try [lua-batteries](https://github.com/clitic/lua-batteries) (created by me) which includes latest version of lua and luarocks. I have compiled fltk4lua with msvc you can check that out from [here](https://github.com/luarocks/luarocks/files/10072914/lua5.4.4%2Bfltk4lua-x86_64-msvc.zip) or else follow these steps.

1. Install C++ build tools (MSVC) from [visual studio](https://visualstudio.microsoft.com).
2. Now download fltk latest sources from [fltk.org](https://www.fltk.org/software.php) and then build fltk with [cmake](https://cmake.org/download) (generator can be any).

```bat
cmake -Bbuild -GNinja
cmake --build build
cmake --install build --prefix d:/fltk
```

3. Finally install fltk4lua with luarocks.

```bat
luarocks install fltk4lua FLTK_DIR=d:/fltk
```

4. Run examples

![Screenshot 2022-11-23 101527](https://user-images.githubusercontent.com/71246790/203473373-0eba9da7-8d2d-44e6-9706-872eece4fa81.png)

**Comment by DiegoJArg** [comment]
Hi @clitic, Your bundle works !! Thank you !!
You should include a link to your bundle at in https://luabinaries.sourceforge.net
It would had helped me a lot.

**Comment by DiegoJArg** [comment]
Thanks to you confirming compiling is posible, I got inspired and achieved compilation with MinGW.

I realised that working with Lua will require quite a lot of C compiling.
I already had MSYS2 with MinGW-w64-i686 (32 bits version) installed.
I picked 32 bit version since it will get less footprint for small applications as compared to 64 bits and I don't attempt to open huge files.

My steps were:

First install MSYS2 and open MSYS2-MINGW32
```
$ pacman -Syu
$ pacman -Syu base-devel
$ pacman -Syu mingw-w64-i686-toolchain
$ pacman -Syu mingw-w64-x86_64-toolchain (64 bits)
$ pacman -Syu mingw-w64-i686-cmake
$ pacman -Syu autoconf
$ pacman -Syu automake
```

Add to the PATH at the first:
C:\msys64\mingw32\bin
C:\msys64\usr\bin
Copy&rename mingw32-make to make
Not placing the paths at the first, may cause dlls from others applications to interfere. It happened to me with zlib1.dll, libpng16-16.dll and installed LuaForWindows.


```
mkdir my_compile_folder
cd my_compile_folder
git clone https://github.com/siffiejoe/fltk
git clone https://github.com/siffiejoe/lua-fltk4lua
git clone https://github.com/siffiejoe/lua-moon lua-fltk4lua/moon
git clone https://github.com/lunarmodules/lua-compat-5.3 compat-5.3
```

```
cd fltk
NOCONFIGURE=1 ./autogen.sh
./configure
make
test/demo
cd ..
cd lua-fltk4lua

<<<<<<EDIT Makefile>>>>>>>
   DLL_INSTALL_DIR = ../fltk/lib
   LUA_INCDIR = ../fltk/FL
   FLTK_CONFIG = ../fltk/fltk-config
   MYLDFLAGS = `${FLTK_CONFIG} --use-images --ldflags` -lstdc++ -llua
   LIB_EXTENSION = dll

make depend
make (this produces a lot of warnings about cast for callback functions to be ignored)
./examples/chart.lua
```

NOTE:
1. After running the example chart.lua (or any other), it appears that Lua doesn't stops and I had to kill it. I didn't inspected much why.
2. The MinGW version got 2.2Mb, but linked libraries seem to be much less. This probably improves nothing.

[my_lua5.4.4+fltk4lua-mingw64-i686.zip](https://github.com/luarocks/luarocks/files/10075991/my_lua5.4.4%2Bfltk4lua-mingw64-i686.zip)

![image](https://user-images.githubusercontent.com/25246225/203561608-37de9da1-d506-4ec1-9616-e37c44cfe5ce.png)
